examples/bip-340/src/bip-340.rs

```diff
@@ -0,0 +1,169 @@
+    let three = FieldElement::TWO() + FieldElement::ONE();
+    if p1_ != p2_ {
+        (y(p2_) - y(p1_)) * (x(p2_) - x(p1_)).pow_self(FieldElement::ZERO() - FieldElement::TWO())
+    } else {
+        three
+            * x(p1_)
+            * x(p1_)
+            * (FieldElement::TWO() * y(p1_)).pow_self(FieldElement::ZERO() - FieldElement::TWO())
+    }
+}
+
+fn point_add(p1: Point, p2: Point) -> Point {
+    // TODO: I tried writing this as an if/else if/else case distinction
+    // (without res) but hacspec doesn't let me do that.
+    let mut res = Option::<FinitePoint>::None;
+    if p1.is_none() {
+        res = p2
+    } else {
+        if p2.is_none() {
+            res = p1
+        } else {
+            let p1_ = p1.unwrap();
+            let p2_ = p2.unwrap();
+            if !((x(p1_) == x(p2_)) && (y(p1_) != y(p2_))) {
+                let lam = compute_lam(p1_, p2_);
+                let x3 = lam * lam - x(p1_) - x(p2_);
+                res = Some((x3, (lam * (x(p1_) - x3) - y(p1_))));
+            }
+        }
+    }
+    res
+}
+
+fn point_mul(s: Scalar, p: Point) -> Point {
+    let mut p = p;
+    let mut q = Option::<FinitePoint>::None;
+    for i in 0..256 {
+        if s.bit(i) {
+            q = point_add(q, p.clone());
+        }
+        // TODO: this requires a clone for some reason. ed25519.rs has the same
+        // function but doesn't require a clone. Is this because our Points are
+        // Options?
+        p = point_add(p.clone(), p.clone());
+    }
+    q
+}
+
+// TODO: I want to remove this because we have point_mul already. But I haven't
+// figured out how to define a global base point G.
+fn point_mul_base(s: Scalar) -> Point {
+    let x = FieldElement::from_byte_seq_be(&G.slice(0, 32));
+    let y = FieldElement::from_byte_seq_be(&G.slice(32, 32));
+    let g: FinitePoint = (x, y);
+    point_mul(s, Some(g))
+}
+
+pub fn bytes_from_point(p: FinitePoint) -> PublicKey {
+    let (x, _) = p;
+    /* TODO: explain what it does */
+    PublicKey::from_slice(&x.to_byte_seq_be(), 0, 32)
+}
+
+fn bytes_to_scalar(b: &SerializedScalar) -> Option<Scalar> {
+    let s = BigInteger::from_byte_seq_be(b);
+    let curve_order = BigInteger::from_byte_seq_be(&Scalar::max_val().to_byte_seq_be());
+    if s > curve_order {
+        Option::<Scalar>::None
+    } else {
+        Option::<Scalar>::Some(Scalar::from_byte_seq_be(b))
+    }
+}
+
+fn bytes_to_seckey_scalar(b: &SerializedScalar) -> Option<Scalar> {
+    let s = bytes_to_scalar(b)?;
+    if s == Scalar::ZERO() {
+        Option::<Scalar>::None
+    } else {
+        Option::<Scalar>::Some(s)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    InvalidSecretKey,
+}
+
+pub type PubkeyGenResult = Result<PublicKey, Error>;
+
+pub fn pubkey_gen(seckey: SecretKey) -> PubkeyGenResult {
+    let d0 = bytes_to_seckey_scalar(&seckey);
+    if d0.is_none() {
+        // TODO: these errors pretty hard to read, even if we define an extra result type
```

Comment:
How would you like them to look like?
One could use constants instead of enums for the error value to shorten it a little (`PubkeyGenResult::Err(INVALID_SECRET_KEY)?;`). But then you can't use nice enums for errors...

Reply:
Would be nice to be able to skip the `PubkeyGenResult::` prefix (as in regular rust).

Reply:
I agree! This is mostly a typechecker issue. There's not general reason this shouldn't be allowed.